hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/ha/SCMHAManagerImplTest.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.hadoop.hdds.utils.db.BatchOperation;
+import org.apache.hadoop.hdds.utils.db.DBStore;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.ozone.test.GenericTestUtils;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.UUID;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+
+/**
+ * Test cases to verify {@link org.apache.hadoop.hdds.scm.ha.SCMHAManagerImpl}.
+ */
+class SCMHAManagerImplTest {
+
+  private String storageBaseDir;
+  private String clusterID;
+  private SCMHAManager primarySCMHAManager;
+
+  @BeforeEach
+  public void setup() throws IOException, InterruptedException {
+    storageBaseDir = GenericTestUtils.getTempPath(
+        SCMHAManagerImplTest.class.getSimpleName() + "-" +
+            UUID.randomUUID());
+    clusterID = UUID.randomUUID().toString();
+    OzoneConfiguration conf = getConfig("scm1", 9894);
+    final StorageContainerManager scm = getMockStorageContainerManager(conf);
+    SCMRatisServerImpl.initialize(clusterID, scm.getScmId(),
+        scm.getScmNodeDetails(), conf);
+    scm.getScmHAManager().start();
+    primarySCMHAManager = scm.getScmHAManager();
+    // Wait for Ratis Server to be ready
+    Thread.sleep(5000);
+  }
+
+  private OzoneConfiguration getConfig(String scmId, int ratisPort) {
+    OzoneConfiguration conf = new OzoneConfiguration();
+    conf.set(ScmConfigKeys.OZONE_SCM_HA_RATIS_STORAGE_DIR, storageBaseDir
+        + File.separator + scmId + File.separator + "ratis");
+    conf.set(HddsConfigKeys.OZONE_METADATA_DIRS, storageBaseDir
+        + File.separator + scmId + File.separator + "metadata");
+    conf.set(ScmConfigKeys.OZONE_SCM_RATIS_PORT_KEY, String.valueOf(ratisPort));
+    return conf;
+  }
+
+  @AfterEach
+  public void cleanup() throws IOException {
+    primarySCMHAManager.stop();
+    FileUtils.deleteDirectory(new File(storageBaseDir));
+  }
+
+  @Test
+  public void testAddSCM() throws IOException, InterruptedException {
+    Assertions.assertEquals(1, primarySCMHAManager.getRatisServer()
+        .getDivision().getGroup().getPeers().size());
+
+    final StorageContainerManager scm2 = getMockStorageContainerManager(
+        getConfig("scm2", 9898));
+    scm2.getScmHAManager().getRatisServer().start();
+    // Wait for Ratis Server to be ready
+    Thread.sleep(5000);
+    final AddSCMRequest request = new AddSCMRequest(
+        clusterID, scm2.getScmId(),
+        "localhost:" + scm2.getScmHAManager().getRatisServer().getDivision()
+            .getRaftServer().getServerRpc().getInetSocketAddress().getPort());
+    primarySCMHAManager.addSCM(request);
+    Assertions.assertEquals(2, primarySCMHAManager.getRatisServer()
+        .getDivision().getGroup().getPeers().size());
+    scm2.getScmHAManager().getRatisServer().stop();
```

Comment:
Please use try to capture the possible exception from addSCM, and close the ratis server in this case.